feat: Support user-defined log-event boundaries to handle logs with mid-line timestamps (fixes #175). (#199)
y-scope/log-surgeon: docs/schema.md

Reviewer: Since we moved away from the `<name>` syntax in the header syntax, we should probably move away from it here too.

The suggestion is to match my suggestions below, but feel free to come up with something better.
(I can't combine the suggestions as the ``` breaks markdown.)

```suggestion
delimiters:_CHARACTERS_
```

(Not sure if the suggestion works since the diff seems bugged.)

Author: Updated.

Reviewer: I didn't notice before, but github seems to highlight certain characters in `txt` code blocks. So actually `$` and `*` are red randomly (https://github.com/SharafMohamed/log-surgeon/blob/log-event-boundary-new/docs/schema.md). This is also in contrast with inline code not having any highlighting.

`$` is extra weird since the first occurrence is red, but not highlighted, while following occurrences are both red and highlighted red.

I tried to look into it and it seems that `txt` isn't a valid/real language and leads to github trying to automatically guess the language.

https://docs.github.com/en/pages/setting-up-a-github-pages-site-with-jekyll/about-github-pages-and-jekyll#syntax-highlighting:
> By default, code blocks on your site will be highlighted by Jekyll. Jekyll uses the [Rouge](https://github.com/rouge-ruby/rouge) highlighter ....
> Rouge only recognizes lower-case language identifiers for fenced code blocks. For a list of supported languages, see [Languages](https://rouge-ruby.github.io/docs/file.Languages.html).

https://rouge-ruby.github.io/docs/file.Languages.html:
There is `plaintext` but no `txt` language.

For example:
` ```txt ` results in:

```txt
$TEST$$ .*$
```

` ```plaintext ` results in:

```plaintext
$TEST$$ .*$
```

Unfortunately, I can't make suggestions to fix since many cases of `txt` are old.

* [Variables](#variables): Defines patterns for capturing specific pieces of the log.
* [Delimiters](#delimiters): Specifies the characters that separate tokens in the log.
* [Headers](#headers): Identifies the boundary between log events. Headers are also treated as
  variables.
  * The first [timestamp](#headers) captured within a header is considered the log event's

>     timestamp.

Author: This is wrong, the left-most is not used. The first matched is used. Consider alternations, or 0-repetitions, where the left-most isn't matched.

**Syntax:**

```txt
delimiters:$CHARACTERS$
```

* `delimiters` is a reserved name for this rule.
* `$CHARACTERS$` is a set of characters that should be treated as delimiters. These characters
  define the boundaries between tokens in the log.

**Example:**
```txt
delimiters: \t\r\n:,!;%

> ```

Author: Added.

* A schema file must contain at least one `delimiters` rule. If multiple `delimiters` rules are
  specified, only the last one will be used.

### Headers

**Syntax:**

```txt
header:Prefix (?<timestamp>[TIMESTAMP-PATTERN]) suffix
```

* Multiple headers can be specified within a schema.

> * Prefix/suffix can be empty, contain static text, or capture other variables if needed.

Author: Updated.

* The timestamp capture can be omitted if the log-event boundary does not contain a timestamp.
* Multiple timestamp captures are allowed within a header. These can exist within regex repetitions
  or alternations.
  * If no timestamps are parsed, the event's logtype has no timestamp.
  * If one or more timestamps are parsed, the event's logtype uses the first timestamp.
* `timestamp` is a reserved name for the capture within a header rule.

> * `[TIMESTAMP-PATTERN]` is a regular expression using the supported

Reviewer: ```suggestion
* `_PREFIX_`, `_SUFFIX_`, and `_TIMESTAMP_PATTERN_` are regular expression using the supported
```

Author: Updated.
